Buildings/Controls/OBC/Utilities/PIDWithInputGains.mo

```diff
@@ -0,0 +1,749 @@
+    "Boolean flag to enable integral action"
+    annotation (Evaluate=true,HideResult=true);
+  final parameter Boolean with_D=controllerType == Buildings.Controls.OBC.CDL.Types.SimpleController.PD or controllerType == Buildings.Controls.OBC.CDL.Types.SimpleController.PID
+    "Boolean flag to enable derivative action"
+    annotation (Evaluate=true,HideResult=true);
+  Buildings.Controls.OBC.CDL.Continuous.Sources.Constant Dzero(
+    final k=0) if not with_D
+    "Zero input signal"
+    annotation (Evaluate=true,HideResult=true,Placement(transformation(extent={{-20,110},{0,130}})));
+  Buildings.Controls.OBC.CDL.Continuous.Sources.Constant Izero(
+    final k=0) if not with_I
+    "Zero input signal"
+    annotation (Placement(transformation(extent={{20,74},{40,94}})));
+  Buildings.Controls.OBC.CDL.Continuous.MultiplyByParameter uS_revAct(
+    final k=revAct/r)
+    "Set point multiplied by reverse action sign"
+    annotation (Placement(transformation(extent={{-200,30},{-180,50}})));
+  Buildings.Controls.OBC.CDL.Continuous.MultiplyByParameter uMea_revAct(
+    final k=revAct/r)
+    "Set point multiplied by reverse action sign"
+    annotation (Placement(transformation(extent={{-180,-50},{-160,-30}})));
+  Buildings.Controls.OBC.CDL.Continuous.Add addPD
+    "Outputs P and D gains added"
+    annotation (Placement(transformation(extent={{20,116},{40,136}})));
+  Buildings.Controls.OBC.CDL.Continuous.Add addPID
+    "Outputs P, I and D gains added"
+    annotation (Placement(transformation(extent={{80,80},{100,100}})));
+  Buildings.Controls.OBC.CDL.Continuous.Subtract antWinErr if with_I
+    "Error for anti-windup compensation"
+    annotation (Placement(transformation(extent={{162,50},{182,70}})));
+  Buildings.Controls.OBC.CDL.Continuous.MultiplyByParameter antWinGai1(k=1/Ni)
+    if with_I "Gain for anti-windup compensation without the proportional gain"
+    annotation (Placement(transformation(extent={{180,-30},{160,-10}})));
+  Buildings.Controls.OBC.CDL.Continuous.Sources.Constant yResSig(
+    final k=y_reset) if with_I
+    "Signal for y_reset"
+    annotation (Placement(transformation(extent={{-140,-90},{-120,-70}})));
+  Buildings.Controls.OBC.CDL.Continuous.Subtract addRes if with_I
+    "Adder for integrator reset"
+    annotation (Placement(transformation(extent={{-100,-90},{-80,-70}})));
+  Buildings.Controls.OBC.CDL.Logical.Sources.Constant cheYMinMax(
+    final k=yMin < yMax)
+    "Check for values of yMin and yMax"
+    annotation (Placement(transformation(extent={{120,-160},{140,-140}})));
+  Buildings.Controls.OBC.CDL.Utilities.Assert assMesYMinMax(
+    message="LimPID: Limits must be yMin < yMax")
+    "Assertion on yMin and yMax"
+    annotation (Placement(transformation(extent={{160,-160},{180,-140}})));
+
+  block Derivative
+    "Block that approximates the derivative of the input"
+    parameter Real y_start=0
+      "Initial value of output (= state)"
+      annotation (Dialog(group="Initialization"));
```

Comment:
CDL does not allow the declaration of a block inside a *composite block* (https://obc.lbl.gov/specification/cdl.html#sec-com-blo).
You will need to implement the `Derivative` block using CDL, and store it for example in `OBC.Utilities.BaseClasses.Derivative`. To simplify the implementation you can assume `k*Td > 1E-17`, and guard against it with a `CDL.Utilities.Assert` block.

Reply:
As suggested, I created OBC.Utilities.BaseClasses.Derivative. I also added <code>CDL.Utilities.Assert</code> to <code>OBC.Utilities.PIDWithInputGains</code> to prevent <code>k*Td <= 1E-17</code>